vim.fs.ext() returns file extension (was: vim.fs.suffix())

yochem opened the pull request:
Problem:

Checking the extension of a file is done often, e.g. in Nvim's codebase for differentiating Lua and Vimscript files in the runtime. The current way to do this in Lua is (1) a Lua pattern match, which has pitfalls such as not considering filenames starting with a period, or (2) `fnamemodify()` which is both hard to discover and hard to use / read if not very familiar with the possible modifiers.

In summary:
- Extracting the extension of a filename is used often, for example to find the type of a file 
- `vim.fs` is incomplete. It is cumbersome and not ergonomic to search for functionality which one would expect to be in the `vim.fs` module.
- `vim.fs.suffix` is more discoverable than `fnamemodify` with the `':e'` argument
- `vim.fs.suffix('test.lua')` is more readable than `vim.fn.fnamemodify("test.lua", ":e")`, especially for users without Vim(script) experience/knowledge.

Solution:

`vim.fs.suffix()` returns the file extension including the leading period of the extension. Similar to the "file suffix" implementation of many other stdlibs (including `fnamemodify(file, ":e")`), a leading period doesn't indicate the start of the suffix. E.g.: the `.git` folder in a repository doesn't have the suffix `.git`, but it simply has no suffix, similar to a folder named `git` or any other filename without periods. Comparison of this behaviour with other standard libs in https://github.com/neovim/neovim/pull/36997#issuecomment-3664944547.

Possible follow-up: add `all` option to return all suffixes:
```lua
vim.fs.suffix('repo.tar.gz', { all = true })
-- { '.tar', '.gz' }
```

justinmk commented:
> Should I also change it for the other functions (`abspath`, `dirname`, and `basename`)?

If possible, though separate PR ideally. Not sure about `abspath`, I think the details were tricky there.

yochem commented:
> I don't think "the filename" excludes the extension.

Sure, call it the "stem" of the filename. 

Here's an overview of how other stdlibs handle it:

| Language | Function | Output |
|----------|----------|--------|
| Vimscript | `fnamemodify('.nvimrc', ':e')` | `''` |
| Node | `path.extname('.nvimrc')` | `''` |
| Ruby | `File.extname('.nvimrc')` | `''` |
| Python | `Path('.nvimrc').suffix` | `''` |
| PHP | `pathinfo(".nvimrc", PATHINFO_EXTENSION)` | `'.nvimrc'` |
| Go | `filepath.Ext(".nvimrc")` | `'.nvimrc'` |
| Rust | `Path::new(".nvimrc").extension()` | `None` |
| Swift | `URL(fileURLWithPath: ".nvimrc").pathExtension` | `''` |
| Kotlin | `File(".nvimrc").extension` | `'nvimrc'` |
| C++ | `std::filesystem::path(".nvimrc").extension()` | `""` |
| Julia | `splitext(".nvimrc")[2]` | `''` |
| Zig | `std.fs.path.extension(".nvimrc")` | `{ }` |

yochem commented:
> why do not just using vim.fn.fnamemodify(file, ':e'), What's the point of adding such wraps besides making the code execute a few more lines.

It's all about the ergonomics of consuming the Nvim interface.

A major advantage of the Lua stdlib is that Nvim's interface can improve without breaking the existing Vimscript interface. The Lua interface is more user friendly, largely because of Lua itself as language, enabling features such as proper namespacing and optional dictionary arguments. This allows for a more modern approach to designing a standard library.

Currently, the stdlib is somewhere in between having only half of functionality. The distinctions between `vim.*`, `vim.api`, `vim.cmd`, and `vim.fn` are clear to users with Vim experience, but they are confusing to newcomers, and I have seen people struggle to understand that difference. Having an `fs` module that only implements part of the functionality one would expect, while related functionality lives in `vim.fn.*` and behaves differently, reduces overall ergonomics rather than improving it. Another advantage of moving this interface into Lua is that the implementation can be changed later on if necessary (so not wrapping `fnamemodify` anymore). Lifting functionality to Lua definitionally fits multiple [goals of the project](https://neovim.io/charter/).

We already have `abspath()`, `basename()`, and `dirname()` in `vim.fs`. A great interface is one where you can guess the function name based on the functionality you need. Not one where you have to endlessly search through the docs to find out that the function you need is _actually a Vimscript function_ (implementation detail) and defined elsewhere.

justinmk reviewed (approved):
LGTM, please rebase? 

Although we generally don't want trivial wrappers duplicating vimscript functions, for `vim.fs` we clearly aligned to have a "complete" interface, and I didn't see any objections to `vim.fs.basename` / `vim.fs.dirname` ...

p.s. @yochem if you plan to change basename/dirname to also use `fnamemodify`, that could clear up stuff like https://github.com/neovim/neovim/pull/38085 , which would be valuable.

yochem committed:
feat(vim.fs): add vim.fs.ext()

Problem:
Checking the extension of a file is done often, e.g. in Nvim's codebase
for differentiating Lua and Vimscript files in the runtime. The current
way to do this in Lua is (1) a Lua pattern match, which has pitfalls
such as not considering filenames starting with a period, or (2)
fnamemodify() which is both hard to discover and hard to use / read if
not very familiar with the possible modifiers.

vim.fs.ext() returns the file extension including the leading period of
the extension. Similar to the "file extension" implementation of many
other stdlibs (including fnamemodify(file, ":e")), a leading period
doesn't indicate the start of the extension. E.g.: the .git folder in a
repository doesn't have the extension .git, but it simply has no
extension, similar to a folder named git or any other filename without
periods.
